llvm/lib/Target/MSP430/MSP430FrameLowering.cpp

```diff
@@ -212,6 +237,12 @@ bool MSP430FrameLowering::restoreCalleeSavedRegisters(
   MachineFunction &MF = *MBB.getParent();
   const TargetInstrInfo &TII = *MF.getSubtarget().getInstrInfo();
 
+  const char *RestorerName = tailCallNameForRestoringCalleSavedRegisters(CSI);
```

Comment:
We have problem here. The RET after JUMP is still on place and nothing will erase it. Or maybe the branch folder does this already for us?

Reply:
> We have problem here. The RET after JUMP is still on place and nothing will erase it.

Exactly. This is what I TODO'ed as "get rid of the dangling ret at the very end of a function" (strictly speaking, not only function but every terminating BB, if I get it right). Right now I'm trying to drop it, but just calling

```cpp
MBB.getLastNonDebugInstr()->removeFromParent();
```

generates an assertion slightly later:

```
Command Output (stderr):
--
Can only insert epilog into returning blocks
UNREACHABLE executed at /home/ast/projects/llvm-project/llvm/lib/Target/MSP430/MSP430FrameLowering.cpp:121!
...
 #7 0x000056363ad77aa9 bindingsErrorHandler(void*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool) /home/ast/projects/llvm-project/llvm/lib/Support/ErrorHandling.cpp:219:55
 #8 0x0000563639470068 llvm::MSP430FrameLowering::emitEpilogue(llvm::MachineFunction&, llvm::MachineBasicBlock&) const /home/ast/projects/llvm-project/llvm/lib/Target/MSP430/MSP430FrameLowering.cpp:119:22
 #9 0x000056363a0c186b (anonymous namespace)::PEI::insertPrologEpilogCode(llvm::MachineFunction&) /home/ast/projects/llvm-project/llvm/lib/CodeGen/PrologEpilogInserter.cpp:1100:3
#10 0x000056363a0bdc71 (anonymous namespace)::PEI::runOnMachineFunction(llvm::MachineFunction&) /home/ast/projects/llvm-project/llvm/lib/CodeGen/PrologEpilogInserter.cpp:259:18
```

Another thing that seems fragile to me (even if It would be appropriate to refactor that code to check *before* emitting epilog) is that I'm not sure non-lazy instruction removal would not invalidate some iterators up the stack. Is there some delayed removal machinery like `setIsDead()` on operand?

Reply:
How about the following: introduce the special pseudo-instruction that would be terminator. Let's call it EPILOGUE. It would have single imm argument – the number of registers. Emit the instruction here. And just during the pseudo instruction expansion erase the RET that would be after it.